### PR TITLE
mgr/cephadm: add support for osd_id_claims

### PR DESCRIPTION
**Pull request by jschmid1**
Signed-off-by: Joshua Schmid <jschmid@suse.de>

Fixes: https://tracker.ceph.com/issues/44769
~Blocked by: https://tracker.ceph.com/issues/44879~
invalid

Todo:
- [x] rebase to master
- [x] decide what to do with the ["HORRIBLE_HACK"](https://github.com/ceph/ceph/blob/5b39c9a2ed0ee82488001a87df5dd20a441b8bff/src/python-common/ceph/deployment/translate.py#L53)

Is the ceph-volume part where `lvm batch` couldn't handle lv/vgs fixed? edit: no
edit: this needs to stay in for now :/
- [ ] Add teuthology tests for osd replacement

### CODE REVIEW

File: src/pybind/mgr/cephadm/module.py [hunk 1]
```diff
@@ -2066,28 +2066,58 @@ def blink(host, dev, path):
         return blink(locs)
 
     def get_osd_uuid_map(self, only_up=False):
-        # type: (bool) -> Dict[str,str]
+        # type: (Optional[bool]) -> Dict[str, str]
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py:2069]: `Optional[bool]` looks always a bit odd to me. If possible, I'd keep it as a simple `bool`

File: src/pybind/mgr/cephadm/module.py [hunk 2]
```diff
@@ -2172,16 +2202,16 @@ def _create_osd(self, host, cmd):
...
-                if osd_id in before_osd_uuid_map:
-                    # this osd existed before we ran prepare
+                if osd_id in before_osd_uuid_map and osd_id not in replace_osd_ids:
+                    # if it exists but is part of the replacement operation, don't skip
                     continue
                 if osd_id not in osd_uuid_map:
                     self.log.debug('osd id %d does not exist in cluster' % osd_id)
                     continue
-                if osd_uuid_map[osd_id] != osd['tags']['ceph.osd_fsid']:
+                if osd_uuid_map.get(osd_id) != osd['tags']['ceph.osd_fsid']:
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py:2211]: `osd_uuid_map.get(osd_id)` is basically the same as `osd_uuid_map.get[osd_id]`. Maybe

```suggestion
                if osd_uuid_map.get(osd_id, None) != osd['tags']['ceph.osd_fsid']:
```
?

File: src/pybind/mgr/cephadm/module.py [hunk 3]
```diff
@@ -2172,16 +2202,16 @@ def _create_osd(self, host, cmd):
...
                     self.log.debug('mismatched osd uuid (cluster has %s, osd '
                                    'has %s)' % (
-                                       osd_uuid_map[osd_id],
+                                       osd_uuid_map.get(osd_id),
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py:2214]: same

File: src/pybind/mgr/cephadm/module.py [hunk 4]
```diff
@@ -2272,7 +2302,7 @@ def _create_daemon(self, daemon_type, daemon_id, host,
             if daemon_type == 'osd':
                 if not osd_uuid_map:
                     osd_uuid_map = self.get_osd_uuid_map()
-                osd_uuid = osd_uuid_map.get(daemon_id, None)
+                osd_uuid = osd_uuid_map.get(daemon_id)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py:2305]: (this will now raise a `KeyError`, if it is missing)

File: src/pybind/mgr/cephadm/module.py (hunk 2 above)
Comment (jschmid1) [reply src/pybind/mgr/cephadm/module.py:2211]: It's not quite the same:

while

```
In [3]: {}['foo']                                                               
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-3-e91fc3e9cce0> in <module>
----> 1 {}['foo']

KeyError: 'foo'
```

raises a KeyError,

the .get() method for dictionaries returns `None` when the key was not found. 

``` python
In [2]: {}.get('foo') is None                                                   
Out[2]: True
```

Hence, 

```python
{}.get('foo', None)
```

is redundant

File: src/pybind/mgr/cephadm/module.py (hunk 4 above)
Comment (jschmid1) [reply src/pybind/mgr/cephadm/module.py:2305]: no, see https://github.com/ceph/ceph/pull/34346#discussion_r402281397

File: src/pybind/mgr/cephadm/module.py (hunk 1 above)
Comment (jschmid1) [reply src/pybind/mgr/cephadm/module.py]: resolved

**Comment by jschmid1** [comment]
jenkins test make check

File: src/python-common/ceph/deployment/translate.py [hunk 5]
```diff
@@ -15,11 +15,13 @@ class to_ceph_volume(object):
 
     def __init__(self,
                  spec,  # type: DriveGroupSpec
-                 selection  # type: DriveSelection
+                 selection,  # type: DriveSelection
+                 host  # type: str
```
Comment (sebastian-philipp) [inline src/python-common/ceph/deployment/translate.py:19]: Adding a `host` parameter to `to_ceph_volume()` feels a bit wrong, as `to_ceph_volume()` is supposed to be independent of the host. What about directly passing `reclaimed_ids` as a parameter and then:

```python
to_ceph_volume(
    spec,
    selection,
    self.spec.osd_id_claims.get(self.host, [])
)
```
?

Comment (jschmid1) [reply src/python-common/ceph/deployment/translate.py:19]: agreed

**Comment by jschmid1** [comment]
> please change the base branch from "octopus" to "master"

Already on the list as you can see in the TODO in the PR body.

**Comment by smithfarm** [comment]
> Already on the list as you can see in the TODO in the PR body.

Yes, I saw that. I added this change request to all outstanding octopus PRs that are not backports.

**Comment by sebastian-philipp** [comment]
@jdurgin who do you think could review the osd related parts?

**Comment by jdurgin** [comment]
The osd-related pieces (relying on the destroyed flag) look good to me. I'm excited for osd replacement in teuthology - it's a big missing piece

File: src/python-common/ceph/deployment/translate.py [hunk 6]
```diff
@@ -46,15 +49,17 @@ def run(self):
...
+        # if self.spec.objectstore == 'bluestore' and \
+        #    not self.spec.encrypted and \
+        #    not self.spec.osds_per_device and \
+        #    len(data_devices) == 1 and \
+        #    not db_devices and \
+        #    not wal_devices:
+        #     cmd = "lvm prepare --bluestore --data %s --no-systemd" % (' '.join(data_devices))
+        #     if self.osd_id_claims:
+        #         cmd += " --osd-id {}".format(" ".join(self.osd_id_claims))
+        #     return cmd
```
Comment (sebastian-philipp) [inline src/python-common/ceph/deployment/translate.py:62]: This will likely break Teuthology, right?

Comment (jschmid1) [reply src/python-common/ceph/deployment/translate.py:62]: I'm about to find, teuthology run is about to get scheduled

**Comment by jschmid1** [comment]
all green -> http://pulpito.ceph.com/jschmid-2020-04-15_14:58:48-rados-wip-jschmid1-testing-2020-04-15-1540-distro-basic-smithi/

**Pull request merged**